Kani assert override is more restrictive than standard assert

This code works as a test but fails with Kani:

```rust
#[cfg_attr(test, test)]
#[cfg_attr(kani, kani::proof)]
fn check() {
    let v: bool = true;
    assert!(&v);
}
```

With test: builds successfully, no warnings/errors, test passes.

With kani:

```text
error[E0308]: mismatched types
  --> test_bool_ref.rs:7:13
   |
7  |     assert!(&v);
   |     --------^^-
   |     |       |
   |     |       expected `bool`, found `&bool`
   |     arguments to this function are incorrect
   |
note: function defined here
  --> /home/ubuntu/rmc/library/kani/src/lib.rs:62:14
   |
62 | pub const fn assert(_cond: bool, _msg: &'static str) {
   |              ^^^^^^
```

Interesting. Apparently, this is valid Rust:
```rust
fn main() {
    let v: bool = true;
    if !&v {}
```
```bash
$ rustc neg.rs
$
```
which is what the standard `assert` macro expands to, but it is not valid without the negation:
```rust
fn main() {
    let v: bool = true;
    if &v {}
```
```bash
$ rustc not_neg.rs 
error[E0308]: mismatched types
 --> not_neg.rs:3:8
  |
3 |     if &v { }
  |        ^^ expected `bool`, found `&bool`
  |
help: consider removing the borrow
  |
3 -     if &v { }
3 +     if v { }
  |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```

--- a/library/std/src/lib.rs
+++ b/library/std/src/lib.rs
@@ -44,10 +44,11 @@ pub mod process;
 #[macro_export]
 macro_rules! assert {
     ($cond:expr $(,)?) => {
-        kani::assert($cond, concat!("assertion failed: ", stringify!($cond)));
+        // The double negation is to resolve https://github.com/model-checking/kani/issues/2108
+        kani::assert(!!$cond, concat!("assertion failed: ", stringify!($cond)));
     };
     ($cond:expr, $($arg:tt)+) => {{
-        kani::assert($cond, concat!(stringify!($($arg)+)));
+        kani::assert(!!$cond, concat!(stringify!($($arg)+)));
         // Process the arguments of the assert inside an unreachable block. This
         // is to make sure errors in the arguments (e.g. an unknown variable or
         // an argument that does not implement the Display or Debug traits) are

--- a/tests/kani/Assert/bool_ref.rs
+++ b/tests/kani/Assert/bool_ref.rs
@@ -1,0 +1,12 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This test makes sure Kani handles the valid `assert!(&b)` syntax where `b` is a `bool`
+//! See https://github.com/model-checking/kani/issues/2108 for details.
+
+#[kani::proof]
+fn check_assert_with_reg() {
+    let b1: bool = kani::any();
+    let b2 = b1 || !b1; // true
+    assert!(&b2);
+}

